File: PHP/src/OneLang/One/Transforms/ResolveUnresolvedTypes.php
<?php

namespace OneLang\One\Transforms\ResolveUnresolvedTypes;

use OneLang\One\AstTransformer\AstTransformer;
use OneLang\One\Ast\AstTypes\UnresolvedType;
use OneLang\One\Ast\AstTypes\ClassType;
use OneLang\One\Ast\AstTypes\InterfaceType;
use OneLang\One\Ast\AstTypes\EnumType;
use OneLang\One\Ast\AstTypes\GenericsType;
use OneLang\One\Ast\Types\Class_;
use OneLang\One\Ast\Types\Interface_;
use OneLang\One\Ast\Types\Enum;
use OneLang\One\Ast\Expressions\Expression;
use OneLang\One\Ast\Expressions\UnresolvedNewExpression;
use OneLang\One\Ast\Expressions\NewExpression;
use OneLang\One\Ast\Interfaces\IType;

class ResolveUnresolvedTypes extends AstTransformer {
    function __construct() {
        parent::__construct("ResolveUnresolvedTypes");
        
    }
    
    protected function visitType($type) {
        parent::visitType($type);
        if ($type instanceof UnresolvedType) {
            if ($this->currentInterface !== null && in_array($type->typeName, $this->currentInterface->typeArguments))
                return new GenericsType($type->typeName);
            
            $symbol = $this->currentFile->availableSymbols->get($type->typeName);
            if ($symbol === null) {
                $this->errorMan->throw("Unresolved type '" . $type->typeName . "' was not found in available symbols");
                return $type;
            }
            
            if ($symbol instanceof Class_)
                return new ClassType($symbol, $type->typeArguments);
            else if ($symbol instanceof Interface_)
                return new InterfaceType($symbol, $type->typeArguments);
            else if ($symbol instanceof Enum)
                return new EnumType($symbol);
            else {
                $this->errorMan->throw("Unknown symbol type: " . $symbol);
                return $type;
            }
        }
        else
            return $type;
    }
    
    protected function visitExpression($expr) {
        if ($expr instanceof UnresolvedNewExpression) {
            $clsType = $this->visitType($expr->cls);
            if ($clsType instanceof ClassType) {
                $newExpr = new NewExpression($clsType, $expr->args);
                $newExpr->parentNode = $expr->parentNode;
                parent::visitExpression($newExpr);
                return $newExpr;
            }
            else {
                $this->errorMan->throw("Excepted ClassType, but got " . $clsType);
                return $expr;
            }
        }
        else
            return parent::visitExpression($expr);
    }
}

File: Python/OneLang/One/Transforms/ResolveUnresolvedTypes.py
from onelang_core import *
import OneLang.One.AstTransformer as astTrans
import OneLang.One.Ast.AstTypes as astTypes
import OneLang.One.Ast.Types as types
import OneLang.One.Ast.Expressions as exprs
import OneLang.One.Ast.Interfaces as ints

class ResolveUnresolvedTypes(astTrans.AstTransformer):
    def __init__(self):
        super().__init__("ResolveUnresolvedTypes")
    
    def visit_type(self, type):
        super().visit_type(type)
        if isinstance(type, astTypes.UnresolvedType):
            if self.current_interface != None and type.type_name in self.current_interface.type_arguments:
                return astTypes.GenericsType(type.type_name)
            
            symbol = self.current_file.available_symbols.get(type.type_name)
            if symbol == None:
                self.error_man.throw(f'''Unresolved type \'{type.type_name}\' was not found in available symbols''')
                return type
            
            if isinstance(symbol, types.Class):
                return astTypes.ClassType(symbol, type.type_arguments)
            elif isinstance(symbol, types.Interface):
                return astTypes.InterfaceType(symbol, type.type_arguments)
            elif isinstance(symbol, types.Enum):
                return astTypes.EnumType(symbol)
            else:
                self.error_man.throw(f'''Unknown symbol type: {symbol}''')
                return type
        else:
            return type
    
    def visit_expression(self, expr):
        if isinstance(expr, exprs.UnresolvedNewExpression):
            cls_type = self.visit_type(expr.cls_)
            if isinstance(cls_type, astTypes.ClassType):
                new_expr = exprs.NewExpression(cls_type, expr.args)
                new_expr.parent_node = expr.parent_node
                super().visit_expression(new_expr)
                return new_expr
            else:
                self.error_man.throw(f'''Excepted ClassType, but got {cls_type}''')
                return expr
        else:
            return super().visit_expression(expr)